File: src/neko_core/agent.py
"""The agentic loop.

    while not done and steps < max_steps:
        response = provider.complete(messages, tools=tool_schemas())
        if response.tool_calls:
            for call in response.tool_calls:
                observation = tools.execute(call)      # read / search / write / bash
                messages.append(tool_result(call, observation))
        else:
            done = True                                # model produced a final answer

The ``max_steps`` cap is load-bearing: an agent without one can loop forever and burn
money. Tool results (including errors and denials) are fed back as observations so the
model can adapt rather than crash.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Callable

from .tools import tool_schemas

DEFAULT_SYSTEM_PROMPT = (
    "You are Neko Core, a local-first coding agent. Complete the user's task by calling "
    "tools.\n"
    "Tools: read_file and search are read-only; write_file and bash change the workspace "
    "and require approval.\n"
    "Work in small steps: inspect before you edit, make the smallest change that solves "
    "the task, and verify your work. When the task is done, reply with a short summary and "
    "stop calling tools."
)

# on_event(kind, data): kind in {"tool_call", "tool_result", "final", "max_steps"}.
EventHook = Callable[[str, Any], None]


@dataclass
class Agent:
    provider: Any                      # neko_core.providers.Provider
    tools: Any                         # neko_core.tool_runtime.ToolRegistry
    max_steps: int = 20
    system_prompt: str = DEFAULT_SYSTEM_PROMPT
    on_event: EventHook | None = None
    messages: list[dict] = field(default_factory=list)

    def run(self, instruction: str) -> str:
        """Run the loop until the model is done or max_steps is hit. Returns the final text."""
        if not self.messages:
            self.messages.append({"role": "system", "content": self.system_prompt})
        self.messages.append({"role": "user", "content": instruction})

        for _ in range(self.max_steps):
            response = self.provider.complete(self.messages, tools=tool_schemas())
            content = response.get("content")
            tool_calls = response.get("tool_calls") or []

            if not tool_calls:
                final = content or ""
                self.messages.append({"role": "assistant", "content": final})
                self._emit("final", final)
                return final

            self.messages.append(_assistant_tool_message(content, tool_calls))
            for call in tool_calls:
                self._emit("tool_call", call)
                observation = self.tools.execute(call["name"], call["arguments"])
                self._emit("tool_result", {"call": call, "observation": observation})
                self.messages.append(
                    {
                        "role": "tool",
                        "tool_call_id": call.get("id") or call["name"],
                        "content": observation,
                    }
                )

        self._emit("max_steps", self.max_steps)
        return f"[stopped: reached max_steps={self.max_steps}]"

    def _emit(self, kind: str, data: Any) -> None:
        if self.on_event is not None:
            self.on_event(kind, data)


def _assistant_tool_message(content: str | None, tool_calls: list[dict]) -> dict:
    """Rebuild the OpenAI-format assistant turn so the next request carries the
    tool_calls the model made (ids must match the following tool results)."""
    return {
        "role": "assistant",
        "content": content or "",
        "tool_calls": [
            {
                "id": call.get("id") or call["name"],
                "type": "function",
                "function": {
                    "name": call["name"],
                    "arguments": json.dumps(call.get("arguments") or {}),
                },
            }
            for call in tool_calls
        ],
    }

File: src/agent.ts
/**
 * The agentic loop.
 *
 *   while not done and steps < maxSteps:
 *     res = await provider.complete(messages, toolSchemas())
 *     if res.tool_calls: for each -> observation = await tools.execute(call); append tool result
 *     else: done (final answer)
 *
 * The maxSteps cap is load-bearing: an agent without one can loop forever and burn money.
 * Tool observations (errors + denials) are fed back so the model adapts rather than crash.
 */
import { CostTracker } from "./cost.ts";
import type { DeltaHook, Provider, ToolCall } from "./providers.ts";
import type { ToolRegistry } from "./tool-runtime.ts";

export const DEFAULT_SYSTEM_PROMPT =
  "You are Neko Code, a local-first coding agent. Complete the user's task by calling tools.\n" +
  "Tools: read_file, search, glob, ls are read-only; write_file, edit, bash change the " +
  "workspace and require approval.\n" +
  "Prefer `edit` for small changes (exact unique string replace) over rewriting whole files.\n" +
  "read_file output is prefixed with line numbers for reference only - never include the " +
  "line-number prefix in edit/write content.\n" +
  "For multi-step tasks, call todo_write to plan and track progress (keep exactly one item " +
  "in_progress); update it as you go.\n" +
  "Find code with search/glob before assuming; verify your change by running tests or bash. " +
  "Be concise - no filler.\n" +
  "Work in small steps: inspect before you edit, make the smallest change that solves the task, " +
  "and verify your work. When the task is done, reply with a short summary and stop calling tools.";

// onEvent(kind, data): kind in {"tool_call", "tool_result", "final", "max_steps"}.
export type EventHook = (kind: string, data: any) => void;

export interface AgentOptions {
  provider: Provider;
  tools: ToolRegistry;
  maxSteps?: number;
  systemPrompt?: string;
  onEvent?: EventHook;
  /** When set, assistant content is streamed chunk-by-chunk as it arrives. */
  onDelta?: DeltaHook;
}

export class Agent {
  private readonly provider: Provider;
  private readonly tools: ToolRegistry;
  private readonly maxSteps: number;
  private readonly systemPrompt: string;
  private readonly onEvent?: EventHook;
  private readonly onDelta?: DeltaHook;
  readonly cost = new CostTracker();
  messages: any[] = [];

  constructor(opts: AgentOptions) {
    this.provider = opts.provider;
    this.tools = opts.tools;
    this.maxSteps = opts.maxSteps ?? 20;
    this.systemPrompt = opts.systemPrompt ?? DEFAULT_SYSTEM_PROMPT;
    this.onEvent = opts.onEvent;
    this.onDelta = opts.onDelta;
  }

  /** Append text to the system prompt (used by /skill). Seeds the base prompt if needed. */
  appendSystem(text: string): void {
    const sys = this.messages.find((m) => m.role === "system");
    if (sys) sys.content += "\n\n" + text;
    else this.messages.unshift({ role: "system", content: this.systemPrompt + "\n\n" + text });
  }

  /** Run the loop until the model is done or maxSteps is hit. Returns the final text.
   * Pass an AbortSignal to support Esc-to-interrupt (stops cleanly between/within steps). */
  async run(instruction: string, signal?: AbortSignal): Promise<string> {
    if (!this.messages.length) {
      this.messages.push({ role: "system", content: this.systemPrompt });
    }
    this.messages.push({ role: "user", content: instruction });

    for (let step = 0; step < this.maxSteps; step++) {
      this.emit("step", step + 1);
      if (signal?.aborted) return "[interrupted]";
      let response;
      try {
        response = await this.provider.complete(this.messages, this.tools.schemas(), this.onDelta, signal);
      } catch (error) {
        if (signal?.aborted) return "[interrupted]";
        throw error;
      }
      this.cost.add(response.usage);
      const toolCalls = response.tool_calls ?? [];

      if (!toolCalls.length) {
        const final = response.content ?? "";
        this.messages.push({ role: "assistant", content: final });
        this.emit("final", final);
        return final;
      }

      this.messages.push(assistantToolMessage(response.content, toolCalls));
      for (const call of toolCalls) {
        this.emit("tool_call", call);
        const observation = await this.tools.execute(call.name, call.arguments);
        this.emit("tool_result", { call, observation });
        this.messages.push({
          role: "tool",
          tool_call_id: call.id || call.name,
          content: observation,
        });
      }
    }

    this.emit("max_steps", this.maxSteps);
    return `[stopped: reached max_steps=${this.maxSteps}]`;
  }

  private emit(kind: string, data: any): void {
    this.onEvent?.(kind, data);
  }
}

/** Rebuild the OpenAI-format assistant turn so the next request carries the tool_calls
 * the model made (ids must match the following tool results). */
function assistantToolMessage(content: string | null, toolCalls: ToolCall[]): any {
  return {
    role: "assistant",
    content: content ?? "",
    tool_calls: toolCalls.map((call) => ({
      id: call.id || call.name,
      type: "function",
      function: { name: call.name, arguments: JSON.stringify(call.arguments ?? {}) },
    })),
  };
}
